File: pkg/snowflake/snowflake.go
// Package snowflake 提供 64 位全局唯一 ID 生成器。
//
// Layout: [time:32][node:17][step:15]
// Epoch:  2026-06-11 14:59:25 UTC (1781161165)
//
// 实现为无锁 CAS:把 (lastTime, step) 打包进单个 atomic 状态字
// (布局与 ID 本身一致),Generate 的临界区只有一条 CompareAndSwap。
// 线程安全;同一节点产出严格单调递增。
//
// 容量上界:秒级时间戳 + 15 bit step = 每节点每秒最多 32768 个 ID,
// 超出后阻塞等待下一秒。
package snowflake

import (
	"fmt"
	"sync/atomic"
	"time"
)

const (
	Epoch    uint64 = 1781161165
	NodeBits uint64 = 17
	StepBits uint64 = 15

	timeShift = NodeBits + StepBits
	nodeShift = StepBits
	stepMask  = (1 << StepBits) - 1
	NodeMask  = (1 << NodeBits) - 1
)

// Node 是单一节点的 ID 生成器。线程安全(无锁 CAS)。
type Node struct {
	nodeShifted uint64        // 预移位好的 node 段,免去每次 Generate 重复移位
	state       atomic.Uint64 // 上一次发出的 ID:[lastTime:32][node:17][step:15]
}

// NewNode 创建一个 SnowFlake 生成器。nodeID 超过 17 位会 panic。
func NewNode(nodeID uint64) *Node {
	if nodeID > NodeMask {
		panic(fmt.Sprintf("snowflake: node ID %d exceeds max %d", nodeID, NodeMask))
	}
	n := &Node{nodeShifted: nodeID << nodeShift}
	n.state.Store(n.nodeShifted) // lastTime=0, step=0
	return n
}

// Generate 产生一个全局唯一、严格单调递增的 uint64 ID。
func (n *Node) Generate() uint64 {
	for {
		old := n.state.Load()
		lastTime := old >> timeShift
		now := nowEpoch()

		var next uint64
		switch {
		case now > lastTime:
			// 时钟前进:换新秒,step 归零
			next = now<<timeShift | n.nodeShifted
		case old&stepMask < stepMask:
			// 同一秒(或时钟回拨):继续消费 lastTime 秒剩余的 step 池,
			// 回拨期间无需自旋等待,单调性由 old+1 保证
			next = old + 1
		default:
			// 当前秒 step 耗尽:阻塞等真实时钟越过 lastTime 后重试
			n.waitNextTime(lastTime)
			continue
		}

		// 两条路径均满足 next > old,CAS 成功即独占该 ID:
		// 唯一性与严格单调性同时成立
		if n.state.CompareAndSwap(old, next) {
			return next
		}
		// CAS 失败说明并发竞争,重读状态再来
	}
}

// GenerateInto 一次 CAS 预留一整段 ID 并填满 dst,替代「循环调用 Generate len(dst) 次」。
//
// 与循环调用 Generate 的差别**只在开销,不在语义**:
//   - 每段只做一次 CAS 和一次取时钟,而不是每个 ID 各一次;高并发下还省掉大量 CAS 重试;
//   - 产出严格单调递增,且与同一 Node 上其它 Generate / GenerateInto 调用者互不重复
//     (CAS 成功即独占整段,其它调用者只能从段尾之后继续)。
//
// dst 内相邻 ID **不保证连续**:一批跨越秒边界时时间段会跳变,中间有空洞。
// 调用方只能依赖「严格递增 + 唯一」,不得假设 dst[i+1] == dst[i]+1。
//
// 容量与阻塞行为同 Generate:每 Node 每秒 32768 个上限不变;一批超过当前秒剩余额度时,
// 先填满本秒再阻塞等下一秒继续 —— 与循环调用 Generate 完全一致,不会更差。
//
// len(dst) == 0 直接返回,不改动任何状态。
//
// ⚠️ dst 由调用方分配:本方法不按外部输入决定分配大小,避免把「客户端可控的数量」
// 变成内存 DoS 面(§9.18 写入侧上限 / §16.5 容量边界)。批量入口仍须自行校验数量上限。
func (n *Node) GenerateInto(dst []uint64) {
	filled := 0
	for filled < len(dst) {
		old := n.state.Load()
		lastTime := old >> timeShift
		now := nowEpoch()

		// base = 本段第一个 ID;avail = 本秒还能给出的个数(恒 >= 1,故不会空转)
		var base, avail uint64
		switch {
		case now > lastTime:
			// 时钟前进:换新秒,本段从 step 0 起,整池可用
			base = now<<timeShift | n.nodeShifted
			avail = stepMask + 1
		case old&stepMask < stepMask:
			// 同一秒(或时钟回拨):从 old+1 起,消费 lastTime 秒剩余的 step 池
			base = old + 1
			avail = stepMask - (old & stepMask)
		default:
			// 当前秒 step 耗尽:阻塞等真实时钟越过 lastTime 后重试
			n.waitNextTime(lastTime)
			continue
		}

		take := uint64(len(dst) - filled)
		if take > avail {
			take = avail
		}
		// last 是本段最后一个 ID。两条路径都满足 last >= base > old(新秒时间段更大;
		// 同秒时 base = old+1),且 take <= avail 保证 step 不会溢出进 node 段。
		last := base + take - 1

		if !n.state.CompareAndSwap(old, last) {
			continue // 并发竞争,重读状态再来
		}
		// CAS 成功 ⇒ 独占 [base, last],可安全逐个写出
		for i := uint64(0); i < take; i++ {
			dst[filled] = base + i
			filled++
		}
	}
}

// waitNextTime 阻塞直到真实时钟越过 last(秒级粒度,sleep 1ms 轮询)。
func (n *Node) waitNextTime(last uint64) {
	for nowEpoch() <= last {
		time.Sleep(time.Millisecond)
	}
}

func nowEpoch() uint64 {
	ts := unixNow()
	if ts < int64(Epoch) {
		// 时钟早于 Epoch 时 uint64 减法会下溢出垃圾时间位,
		// 产出的 ID 会与历史 ID 冲突——宁可 panic 也不能发错 ID
		panic(fmt.Sprintf("snowflake: system clock %d is before epoch %d (%s)",
			ts, Epoch, time.Unix(int64(Epoch), 0).UTC().Format(time.RFC3339)))
	}
	return uint64(ts) - Epoch
}

// MinIDAt 返回 unixSec 时刻(Unix 秒)可能生成的最小 ID:时间段取该秒、node/step 全零。
// 用途:按"创建时间早于 cutoff"做范围清理(如 player_mail_claim 按 mail_id < MinIDAt(cutoff)
// 批量删),把时间条件转成 ID 范围条件,走主键/索引扫描。
// unixSec 早于 Epoch 时返回 0(全部 ID 都晚于该时刻)。
func MinIDAt(unixSec int64) uint64 {
	if unixSec < int64(Epoch) {
		return 0
	}
	return (uint64(unixSec) - Epoch) << timeShift
}

// unixNow 返回当前 Unix 秒。抽成包级变量仅为便于测试注入虚拟时钟
// (绕开 32768 ID/s 的真实容量墙做大规模并发验证);生产路径恒等于
// time.Now().Unix()。
var unixNow = func() int64 { return time.Now().Unix() }

File: python/pandorapy/snowflake.py
"""Snowflake 发号器 —— 与 Go 侧 pkg/snowflake 的位布局**完全一致**。

为什么位布局必须逐位一致:
    dialogue_id / trade_id / mail 游标等等都是 snowflake,它们已经落库、已经在客户端
    手里、已经在日志里。Python 版和 Go 版在迁移期会同时发号(strangler 双栈),
    如果位布局不同:
      - 同一个逻辑秒里两边可能发出**相同**的 ID(node 段错位)
      - 或者 ID 大小顺序错乱(时间段位宽不同),破坏"ID 单调递增"这个被依赖的性质
    这是静默数据损坏,不是崩溃。

Go 侧口径(pkg/snowflake/snowflake.go:22):
    Epoch    = 1781161165        # 2026-06-11 06:59:25 UTC,**秒**级,不是毫秒
    NodeBits = 17                # node 段,最多 131072 个节点
    StepBits = 15                # 每个逻辑秒每节点 32768 个序号
    timeShift = 17 + 15 = 32
    nodeShift = 15

    ID = (unix_sec - Epoch) << 32 | (node_id << 15) | step

⚠️ 时间粒度是**秒**而不是毫秒 —— 这点和大多数 snowflake 实现不同,照抄网上的实现
   一定会错。每节点每秒上限 32768 个 ID,超了就阻塞到下一秒(与 Go 一致,不是丢号)。

node_id 唯一性:
    Go 侧有 pkg/snowflake/etcdnode 做 etcd 自动抢占 + 失租退出。Python 侧本轮只实现
    static 模式(读 yaml 的 node.node_id),够 dialogue 单副本用。多副本部署前必须先
    补 etcd 抢占,否则会重号 —— 见 CLAUDE.md §9 不变量 11。
"""

from __future__ import annotations

import threading
import time

# ── 位布局(与 Go 逐个对齐,不要改)────────────────────────────────────────────
EPOCH: int = 1781161165
NODE_BITS: int = 17
STEP_BITS: int = 15

_TIME_SHIFT = NODE_BITS + STEP_BITS  # 32
_NODE_SHIFT = STEP_BITS  # 15
_STEP_MASK = (1 << STEP_BITS) - 1  # 32767
NODE_MASK: int = (1 << NODE_BITS) - 1  # 131071


class ClockBeforeEpochError(RuntimeError):
    """系统时钟早于 Epoch。

    Go 侧在这里直接 panic(snowflake.go:202),理由是:uint64 减法会下溢出成垃圾时间位,
    发出的 ID 会污染全局有序性且无法回收。Python 侧同样必须硬失败而不是返回一个坏 ID。
    """


class Node:
    """一个发号节点。线程安全(对应 Go 侧的 sync.Mutex)。

    注意:这里用线程锁而不是 asyncio.Lock —— 发号是纯 CPU 操作、不 await,
    用线程锁能同时保护 asyncio 与线程池两条调用路径,且不需要把 Generate 变成协程
    (变协程会让每个调用点都得 await,污染整条业务链路)。
    """

    __slots__ = ("_node_shifted", "_last_sec", "_step", "_lock", "node_id")

    def __init__(self, node_id: int) -> None:
        if not 0 <= node_id <= NODE_MASK:
            raise ValueError(
                f"snowflake node_id={node_id} 超出范围 [0, {NODE_MASK}]"
                f"(NodeBits={NODE_BITS});多副本请走 etcd 自动分配"
            )
        self.node_id = node_id
        self._node_shifted = node_id << _NODE_SHIFT
        self._last_sec = -1
        self._step = 0
        self._lock = threading.Lock()

    def generate(self) -> int:
        """铸一个 ID。步池耗尽时阻塞到下一逻辑秒(与 Go 一致:不丢号、不重号)。"""
        with self._lock:
            now = _now_epoch()
            if now == self._last_sec:
                self._step = (self._step + 1) & _STEP_MASK
                if self._step == 0:
                    # 本逻辑秒的 32768 个序号用光了 —— 等到时钟走过这一秒。
                    # Go 侧在这里会打一条 error 日志(超过 2s 视为时钟异常),
                    # Python 侧交由调用方观测 QPS,不在锁内做 I/O。
                    now = _wait_next_second(self._last_sec)
                    self._last_sec = now
            else:
                if now < self._last_sec:
                    # 时钟回拨:继续用上一个逻辑秒,序号往下走。绝不发小于已发出的 ID。
                    self._step = (self._step + 1) & _STEP_MASK
                    if self._step == 0:
                        now = _wait_next_second(self._last_sec)
                        self._last_sec = now
                    return (self._last_sec << _TIME_SHIFT) | self._node_shifted | self._step
                self._last_sec = now
                self._step = 0
            return (self._last_sec << _TIME_SHIFT) | self._node_shifted | self._step


def _now_epoch() -> int:
    """当前时间距 Epoch 的**秒**数。时钟早于 Epoch 直接硬失败。"""
    ts = int(time.time())
    if ts < EPOCH:
        raise ClockBeforeEpochError(
            f"系统时钟 {ts} 早于 snowflake epoch {EPOCH} "
            f"({time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime(EPOCH))})"
        )
    return ts - EPOCH


def _wait_next_second(last_sec: int) -> int:
    """自旋到时钟越过 last_sec。"""
    while True:
        now = _now_epoch()
        if now > last_sec:
            return now
        # 秒级粒度,睡到本秒结束就够了;不做忙等,避免烧 CPU。
        time.sleep(0.005)


def timestamp_of(snowflake_id: int) -> int:
    """从 ID 反解 unix 秒。排障时把一个 ID 还原成"什么时候发的"。"""
    return (snowflake_id >> _TIME_SHIFT) + EPOCH


def node_of(snowflake_id: int) -> int:
    """从 ID 反解 node_id。查"重号是哪两个副本"时用。"""
    return (snowflake_id >> _NODE_SHIFT) & NODE_MASK


def step_of(snowflake_id: int) -> int:
    """从 ID 反解逻辑秒内序号。"""
    return snowflake_id & _STEP_MASK
